File: adventjs/2025/04/solution.ts
function decodeSantaPin(code: string): string | null {
  const blocks = code.match(/\[([^\]]+)\]/g)?.map((m) => m.slice(1, -1)) || []
  if (blocks.length < 4) return null

  const digits: number[] = []
  for (const block of blocks) {
    if (block === "<") {
      if (!digits.length) return null
      digits.push(digits[digits.length - 1])
    } else {
      if (!/^\d/.test(block)) return null
      let digit = parseInt(block[0], 10)
      for (const op of block.slice(1)) {
        digit = op === "+" ? (digit + 1) % 10 : (digit - 1 + 10) % 10
      }
      digits.push(digit)
    }
  }
  return digits.join("")
}

File: adventjs/2025/04/solution.py
import re
from typing import Optional


def decode_santa_pin(code: str) -> Optional[str]:
    blocks = re.findall(r'\[([^\]]+)\]', code)
    if len(blocks) < 4:
        return None

    digits = []
    for block in blocks:
        if block == '<':
            if not digits:
                return None
            digits.append(digits[-1])
            continue
        if not block[0].isdigit():
            return None
        digit = int(block[0])
        for op in block[1:]:
            digit = (digit + 1) % 10 if op == '+' else (digit - 1 + 10) % 10
        digits.append(digit)
    return ''.join(str(d) for d in digits)
